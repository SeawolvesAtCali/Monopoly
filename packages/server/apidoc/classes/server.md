[@prisel/server](../README.md) > [Server](../classes/server.md)

# Class: Server

## Hierarchy

**Server**

## Index

### Methods

* [broadcast](server.md#broadcast)
* [close](server.md#close)
* [emit](server.md#emit)
* [register](server.md#register)
* [start](server.md#start)

---

## Methods

<a id="broadcast"></a>

###  broadcast

▸ **broadcast**(roomId: *`any`*, messageType: *`string`*, data: *`any`*): `void`

*Defined in [server.ts:75](https://github.com/SeawolvesAtCali/prisel/blob/4c45c20/packages/server/server.ts#L75)*

**Parameters:**

| Name | Type |
| ------ | ------ |
| roomId | `any` |
| messageType | `string` |
| data | `any` |

**Returns:** `void`

___
<a id="close"></a>

###  close

▸ **close**(): `void`

*Defined in [server.ts:87](https://github.com/SeawolvesAtCali/prisel/blob/4c45c20/packages/server/server.ts#L87)*

**Returns:** `void`

___
<a id="emit"></a>

###  emit

▸ **emit**(client: *`Socket`*, messageType: *`string`*, data: *`any`*): `void`

*Defined in [server.ts:81](https://github.com/SeawolvesAtCali/prisel/blob/4c45c20/packages/server/server.ts#L81)*

**Parameters:**

| Name | Type |
| ------ | ------ |
| client | `Socket` |
| messageType | `string` |
| data | `any` |

**Returns:** `void`

___
<a id="register"></a>

###  register

▸ **register**(game: *[GameConfig](../#gameconfig)*, room?: *[RoomConfig](../#roomconfig)*): `void`

*Defined in [server.ts:71](https://github.com/SeawolvesAtCali/prisel/blob/4c45c20/packages/server/server.ts#L71)*

**Parameters:**

| Name | Type | Default value |
| ------ | ------ | ------ |
| game | [GameConfig](../#gameconfig) | - |
| `Default value` room | [RoomConfig](../#roomconfig) |  BaseRoomConfig |

**Returns:** `void`

___
<a id="start"></a>

###  start

▸ **start**(): `void`

*Defined in [server.ts:24](https://github.com/SeawolvesAtCali/prisel/blob/4c45c20/packages/server/server.ts#L24)*

**Returns:** `void`

___

