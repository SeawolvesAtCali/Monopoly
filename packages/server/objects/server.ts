import WebSocket from 'ws';

export type Server = WebSocket.Server;
