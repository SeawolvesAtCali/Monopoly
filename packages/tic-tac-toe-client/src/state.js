import { RoomChangePayload, Client, PacketType } from '@prisel/client';
import { useState, useRef, useEffect, useMemo, useCallback } from 'react';

export const phases = {
    LOGIN: 0,
    LOBBY: 1,
    ROOM: 2,
    GAME: 3,
};

const initialRoomState = {
    id: undefined,
    name: undefined,
    host: undefined,
    players: [],
};

export function useConnected(initial = false) {
    const connectedRef = useRef(initial);
    return [connectedRef.current, (newConnected) => (connectedRef.current = newConnected)];
}

export function useRoomState() {
    const [roomId, setRoomId] = useState(undefined);
    const [roomName, setRoomName] = useState(undefined);
    const [players, setPlayers] = useState([]);
    const [host, setHost] = useState(undefined);

    const onJoin = useCallback((roomInfoPayload) => {
        setRoomId(roomInfoPayload.id);
        setRoomName(roomInfoPayload.name);
    }, []);

    const onRoomStateChange = useCallback((roomChangePayload) => {
        if (roomChangePayload.newHost) {
            setHost(roomChangePayload.newHost);
        }
        if (roomChangePayload.newJoins) {
            setPlayers((oldPlayers) => oldPlayers.concat(roomChangePayload.newJoins));
        }
        if (roomChangePayload.newLeaves) {
            setPlayers((oldPlayers) =>
                oldPlayers.filter(
                    (playerInRoom) => !roomChangePayload.newLeaves.includes(playerInRoom),
                ),
            );
        }
    }, []);

    const onLeave = useCallback(() => {
        // clear room state
        setPlayers([]);
        setHost(undefined);
    }, []);

    return { roomId, roomName, players, host, onJoin, onRoomStateChange, onLeave };
}

const initialGameState = {
    winner: undefined,
    currentPlayer: undefined,
    map: [],
    player: [],
};
export function useGameState(client, onEnd) {
    const [gameState, setGameState] = useState(initialGameState);
    const [winner, setWinner] = useState(undefined);
    const endRef = useRef(null);
    endRef.current = onEnd;
    function handleMove(index) {
        const movePacket = {
            type: PacketType.DEFAULT,
            action: 'MESSAGE',
            payload: index,
        };
        client.emit(movePacket);
    }
    useEffect(() => {
        return client.on('GAME_STATE', (packet) => {
            const newState = packet.payload;
            setGameState(newState);
            if (newState.winner) {
                setWinner(newState.winner);
            }
        });
    }, []);

    useEffect(() => {
        return client.on('GAME_OVER', (packet) => {
            setGameState(initialGameState);
            endRef.current(winner);
        });
    }, [winner]);

    return { gameState, handleMove };
}

export function useUserInfo() {
    const [userId, setUserId] = useState(undefined);
    const [username, setUsername] = useState(undefined);
    return {
        id: userId,
        name: username,
        setUserInfo(newId, newName) {
            setUserId(newId);
            setUsername(newName);
        },
    };
}

async function clientSetup(client) {
    return await client.connect();
}
export function useClient(url, connectedRef) {
    const client = useMemo(() => {
        return new Client(url);
    }, []);
    useEffect(() => {
        clientSetup(client).then(() => {
            connectedRef.current = true;
        });
        return () => {
            if (connectedRef.current) {
                console.log('disconnect');
                client.exit();
                connectedRef.current = false;
            }
        };
    }, [client]);
    async function login(username) {
        if (!connectedRef.current) {
            return;
        }
        const loginInfo = await client.login(username);
        if (connectedRef.current) {
            return { id: loginInfo.userId, name };
        }
        throw new Error('disconnected while trying to login');
    }
    return { client, login };
}
