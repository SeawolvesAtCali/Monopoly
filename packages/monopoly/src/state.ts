import { ClientId, Handle } from '@prisel/server';
import Game from './Game';
import { create as createPlayer } from './Player';
import { create as createGame } from './Game';
import createBoard from './gameData/board';
import Node from './Node';
import genId from './genId';

const CASH = 1000;

export function createIntialState(players: ClientId[], handle: Handle): Game {
    const board = createBoard(handle);
    const playerMap = new Map(
        players.map((player) => [
            player,
            createPlayer(
                {
                    cash: CASH,
                    id: player,
                    owning: [],
                    rolled: false,
                    position: board,
                },
                handle,
            ),
        ]),
    );

    const game = createGame(
        {
            id: genId(),
            players: playerMap,
            turnOrder: Array.from(playerMap.values()),
            map: board,
        },
        handle,
    );
    return game;
}

function flattenMap(node: Node): Node[] {
    const nodeSet = new Set<Node>();
    let currentNode = node;
    while (currentNode && !nodeSet.has(currentNode)) {
        nodeSet.add(currentNode);
        currentNode = currentNode.next;
    }
    return Array.from(nodeSet);
}

export function flattenState(game: Game): object {
    return {
        currentPlayer: game.turnOrder[0].id,
        players: Array.from(game.players.values()).map((player) => player.flat()),
        map: flattenMap(game.map).map((node) => ({
            node: node.flat(),
            property: node.property ? node.property.flat() : null,
        })),
    };
}
