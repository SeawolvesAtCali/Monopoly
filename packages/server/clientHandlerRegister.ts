import { Context, Socket } from './objects';
import { MessageType, Payload } from '@prisel/common';

export type Handler = (context: Context, socket: Socket) => (data: Payload) => void;

class ClientHandlerRegister {
    private map = new Map<MessageType, Handler>();
    public push(key: MessageType, handler: Handler) {
        this.map.set(key, handler);
    }

    public get(key: MessageType) {
        return this.map.get(key);
    }

    public get messageList() {
        return this.map.keys();
    }
}

export default new ClientHandlerRegister();
