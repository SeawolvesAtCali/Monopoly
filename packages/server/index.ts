import { Context, Socket } from './objects';
import createContext from './createContext';
import debug from './debug';
import './handler';
import { createServer, watchForDisconnection, getConnectionToken, emit } from './networkUtils';
import clientHandlerRegister, { Handler } from './clientHandlerRegister';
import { parsePacket, RoomType } from '@monopoly/common';
import { handleDisconnect } from './handler/handleDisconnect';
import { getWelcome } from './message/room';

process.stdout.write('starting server');

process.title = 'monopoly-server';

export const server = createServer();
const context: Context = createContext({
    server,
});
// @ts-ignore
// make the context available when debugging in chrome
global.context = context;

server.on('connection', (socket) => {
    debug('client connected');
    emit(socket, ...getWelcome());
    socket.on('message', (data: any) => {
        if (data) {
            const packet = parsePacket(data);
            clientHandlerRegister.forEach(([event, handler]) => {
                if (event === packet.type) {
                    handler(context, socket)(packet.payload);
                }
            });
        }
    });

    const connectionToken = getConnectionToken();
    socket.on('disconnect', () => {
        connectionToken.safeDisconnect();
        handleDisconnect(context, socket);
        debug('client disconnected');
    });

    watchForDisconnection(socket, connectionToken).then(() => {
        if (!connectionToken.safeDisconnected) {
            // client is not responding
            // forcefully terminate connection
            debug(`client ${context.SocketManager.getId(socket)} lost connection`);
            socket.terminate();
            handleDisconnect(context, socket);
        }
    });
});

process.on('exit', () => {
    debug('on exit');
    server.close();
});
process.on('SIGINT', () => {
    debug('on siginit');
    server.close();
    process.exit();
});
