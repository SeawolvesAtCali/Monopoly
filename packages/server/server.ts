import { Context, Socket } from './objects';
import createContext from './createContext';
import debug from './debug';
import './handler';
import {
    createServer,
    watchForDisconnection,
    getConnectionToken,
    emit,
    broadcast,
} from './utils/networkUtils';
import clientHandlerRegister, { Handler } from './clientHandlerRegister';
import { parsePacket } from '@prisel/common';
import { handleDisconnect } from './handler/handleDisconnect';
import { getWelcome } from './message/room';
import { GameConfig } from './utils/gameConfig';
import { RoomConfig, BaseRoomConfig } from './utils/roomConfig';
import ConfigManager from './utils/configManager';

export class Server {
    private context: Context;
    private configManager = new ConfigManager();

    public start() {
        const server = createServer();
        const context: Context = createContext({
            server,
            getConfigs: this.configManager.get.bind(this.configManager),
        });
        this.context = context;

        server.on('connection', (socket) => {
            debug('client connected');
            emit(socket, ...getWelcome());
            socket.on('message', (data: any) => {
                debug(data);
                if (data) {
                    const packet = parsePacket(data);
                    let handled = false;
                    clientHandlerRegister.forEach(([event, handler]) => {
                        if (event === packet.type) {
                            handler(context, socket)(packet.payload);
                            handled = true;
                        }
                    });
                    if (!handled) {
                        debug(`Cannot find handler for ${packet.type}`);
                    }
                }
            });

            const connectionToken = getConnectionToken();
            socket.on('disconnect', () => {
                connectionToken.safeDisconnect();
                handleDisconnect(context, socket)({});
                debug('client disconnected');
            });

            watchForDisconnection(socket, connectionToken).then(() => {
                if (!connectionToken.safeDisconnected) {
                    // client is not responding
                    // forcefully terminate connection
                    debug(`client ${context.SocketManager.getId(socket)} lost connection`);
                    socket.terminate();
                    handleDisconnect(context, socket);
                }
            });
        });
    }

    public register(game: GameConfig, room: RoomConfig = BaseRoomConfig) {
        this.configManager.add(game, room);
    }

    public broadcast(roomId: any, messageType: string, data: any) {
        if (this.context) {
            broadcast(this.context, roomId, messageType, data);
        }
    }

    public emit(client: Socket, messageType: string, data: any) {
        if (this.context) {
            emit(client, messageType, data);
        }
    }

    public close() {
        if (this.context) {
            this.context.server.close();
            this.context = undefined;
        }
    }
}

export default Server;
