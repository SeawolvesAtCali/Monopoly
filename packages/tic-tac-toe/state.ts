import clientHandlerRegister from '@monopoly/server/lib/clientHandlerRegister';
import { Context, Socket, Server } from '@monopoly/server';
import { broadcast } from '@monopoly/server/lib/networkUtils';
import { Messages } from '@monopoly/server';
import { RoomType, GameType } from '@monopoly/common';
import debug from 'debug';
function createGameState() {
    const state: any = {
        player: [undefined, undefined],
        map: ['', '', '', '', '', '', '', '', ''],
        currentPlayer: 0,
    };
    return state;
}

const getRoomId = (context: Context, client: Socket) => {
    const { SocketManager, StateManager } = context;
    const userId = SocketManager.getId(client);
    const roomId = StateManager.connections[userId].roomId;
    return roomId;
};

const handleGameStart = (context: Context, client: Socket) => (data: any) => {
    const state = createGameState();
    const roomId = getRoomId(context, client);
    const { updateState } = context;
    updateState((draftState) => {
        state.player[0] = draftState.rooms[roomId].host;
        state.player[1] = draftState.rooms[roomId].guests[0];
        draftState.rooms[roomId].gameState = state;
    });
    broadcast(context, roomId, ...Messages.getSuccess(RoomType.GAME_START, {}));
    broadcast(
        context,
        roomId,
        ...Messages.getGameState(context.StateManager.rooms[roomId].gameState),
    );
};

export const isEven = (map: string[]) => {
    const hasEmpty = map.some((element: string) => {
        return element === '';
    });
    return !hasEmpty;
};

export const handleMoveImpl = (context: Context, roomId: string, moveData: any) => {
    context.updateState((draftState) => {
        const { gameState } = draftState.rooms[roomId];
        const sign = gameState.currentPlayer === 0 ? 'O' : 'X';
        gameState.map[moveData.index] = sign;
        gameState.currentPlayer = 1 - gameState.currentPlayer;
    });
    return context.StateManager.rooms[roomId].gameState;
};
export const handleMove = (context: Context, client: any) => (data: any) => {
    const { SocketManager, StateManager, updateState } = context;
    const roomId = getRoomId(context, client);
    const state = StateManager.rooms[roomId].gameState;
    const userId = SocketManager.getId(client);
    if (userId !== state.player[state.currentPlayer]) {
        return;
    }
    const newState = handleMoveImpl(context, roomId, data);

    broadcast(context, roomId, ...Messages.getGameState(newState));

    if (checkWin(newState)) {
        process.stdout.write(userId + ' won the game');
        return;
    }

    if (isEven(newState.map)) {
        process.stdout.write('EVEN');
    }
};

export function checkWin(state: any) {
    for (let i = 0; i < 3; i++) {
        if (
            state.map[3 * i] !== '' &&
            state.map[3 * i] === state.map[3 * i + 1] &&
            state.map[3 * i] === state.map[3 * i + 2]
        ) {
            debug('debug')('yes');
            return true;
        }
        if (
            state.map[i] !== '' &&
            state.map[i] === state.map[i + 3] &&
            state.map[i] === state.map[i + 6]
        ) {
            debug('debug')('yes');
            return true;
        }
    }
    if (state.map[0] !== '' && state.map[0] === state.map[4] && state.map[0] === state.map[8]) {
        return true;
    }
    if (state.map[2] !== '' && state.map[2] === state.map[4] && state.map[2] === state.map[6]) {
        return true;
    }
    return false;
}

clientHandlerRegister.push([RoomType.GAME_START, handleGameStart]);
clientHandlerRegister.push([GameType.MOVE, handleMove]);

const server = new Server();
server.start();
