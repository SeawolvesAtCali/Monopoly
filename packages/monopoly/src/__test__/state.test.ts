import { createIntialState } from '../state';
import { newPlayer } from '@prisel/server';
jest.mock('../genId', () => ({
    __esModule: true,
    default: () => 'gen_id',
}));

test('initial state', () => {
    // const mockHandle = {
    //     emit: (playerId: string, data: object) => {
    //         expect(data).toMatchSnapshot();
    //         expect(JSON.stringify(data)).toMatchSnapshot();
    //     },
    //     log: (formatter: string, ...rest: any[]) => {},
    // } as Handle;

    // const game = createIntialState(['player1', 'player2']);
    // game.processMessage(mockHandle, 'player1', { type: 'debug' });
    expect(1).toBe(1);
});
