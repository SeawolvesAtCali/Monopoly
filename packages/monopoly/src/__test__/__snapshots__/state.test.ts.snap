// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`initial state 1`] = `
Object {
  "currentPlayer": "player1",
  "map": Array [
    Object {
      "node": Object {
        "id": "gen_id",
        "next": Object {
          Symbol(refId): "gen_id",
        },
        "prev": Object {
          Symbol(refId): "gen_id",
        },
        "property": null,
      },
      "property": null,
    },
    Object {
      "node": Object {
        "id": "gen_id",
        "next": Object {
          Symbol(refId): "gen_id",
        },
        "prev": Object {
          Symbol(refId): "gen_id",
        },
        "property": Object {
          Symbol(refId): "gen_id",
        },
      },
      "property": Object {
        "id": "gen_id",
        "name": "Medierranean Avenue",
        "owner": null,
        "price": 60,
        "rent": 100,
      },
    },
    Object {
      "node": Object {
        "id": "gen_id",
        "next": Object {
          Symbol(refId): "gen_id",
        },
        "prev": Object {
          Symbol(refId): "gen_id",
        },
        "property": Object {
          Symbol(refId): "gen_id",
        },
      },
      "property": Object {
        "id": "gen_id",
        "name": "Baltic Avenue",
        "owner": null,
        "price": 60,
        "rent": 100,
      },
    },
    Object {
      "node": Object {
        "id": "gen_id",
        "next": Object {
          Symbol(refId): "gen_id",
        },
        "prev": Object {
          Symbol(refId): "gen_id",
        },
        "property": Object {
          Symbol(refId): "gen_id",
        },
      },
      "property": Object {
        "id": "gen_id",
        "name": "Oriental Avenue",
        "owner": null,
        "price": 100,
        "rent": 100,
      },
    },
    Object {
      "node": Object {
        "id": "gen_id",
        "next": Object {
          Symbol(refId): "gen_id",
        },
        "prev": Object {
          Symbol(refId): "gen_id",
        },
        "property": Object {
          Symbol(refId): "gen_id",
        },
      },
      "property": Object {
        "id": "gen_id",
        "name": "Vermont Avenue",
        "owner": null,
        "price": 100,
        "rent": 100,
      },
    },
    Object {
      "node": Object {
        "id": "gen_id",
        "next": Object {
          Symbol(refId): "gen_id",
        },
        "prev": Object {
          Symbol(refId): "gen_id",
        },
        "property": Object {
          Symbol(refId): "gen_id",
        },
      },
      "property": Object {
        "id": "gen_id",
        "name": "Connecticut Avenue",
        "owner": null,
        "price": 120,
        "rent": 100,
      },
    },
    Object {
      "node": Object {
        "id": "gen_id",
        "next": Object {
          Symbol(refId): "gen_id",
        },
        "prev": Object {
          Symbol(refId): "gen_id",
        },
        "property": Object {
          Symbol(refId): "gen_id",
        },
      },
      "property": Object {
        "id": "gen_id",
        "name": "St. Charles Place",
        "owner": null,
        "price": 140,
        "rent": 100,
      },
    },
    Object {
      "node": Object {
        "id": "gen_id",
        "next": Object {
          Symbol(refId): "gen_id",
        },
        "prev": Object {
          Symbol(refId): "gen_id",
        },
        "property": Object {
          Symbol(refId): "gen_id",
        },
      },
      "property": Object {
        "id": "gen_id",
        "name": "States Avenue",
        "owner": null,
        "price": 140,
        "rent": 100,
      },
    },
    Object {
      "node": Object {
        "id": "gen_id",
        "next": Object {
          Symbol(refId): "gen_id",
        },
        "prev": Object {
          Symbol(refId): "gen_id",
        },
        "property": Object {
          Symbol(refId): "gen_id",
        },
      },
      "property": Object {
        "id": "gen_id",
        "name": "Virginia Avenue",
        "owner": null,
        "price": 160,
        "rent": 100,
      },
    },
    Object {
      "node": Object {
        "id": "gen_id",
        "next": Object {
          Symbol(refId): "gen_id",
        },
        "prev": Object {
          Symbol(refId): "gen_id",
        },
        "property": Object {
          Symbol(refId): "gen_id",
        },
      },
      "property": Object {
        "id": "gen_id",
        "name": "St.James Place",
        "owner": null,
        "price": 180,
        "rent": 100,
      },
    },
    Object {
      "node": Object {
        "id": "gen_id",
        "next": Object {
          Symbol(refId): "gen_id",
        },
        "prev": Object {
          Symbol(refId): "gen_id",
        },
        "property": Object {
          Symbol(refId): "gen_id",
        },
      },
      "property": Object {
        "id": "gen_id",
        "name": "Tennessee Avenue",
        "owner": null,
        "price": 180,
        "rent": 100,
      },
    },
    Object {
      "node": Object {
        "id": "gen_id",
        "next": Object {
          Symbol(refId): "gen_id",
        },
        "prev": Object {
          Symbol(refId): "gen_id",
        },
        "property": Object {
          Symbol(refId): "gen_id",
        },
      },
      "property": Object {
        "id": "gen_id",
        "name": "New York Avenue",
        "owner": null,
        "price": 200,
        "rent": 100,
      },
    },
    Object {
      "node": Object {
        "id": "gen_id",
        "next": Object {
          Symbol(refId): "gen_id",
        },
        "prev": Object {
          Symbol(refId): "gen_id",
        },
        "property": Object {
          Symbol(refId): "gen_id",
        },
      },
      "property": Object {
        "id": "gen_id",
        "name": "Kentucky Avenue",
        "owner": null,
        "price": 220,
        "rent": 100,
      },
    },
    Object {
      "node": Object {
        "id": "gen_id",
        "next": Object {
          Symbol(refId): "gen_id",
        },
        "prev": Object {
          Symbol(refId): "gen_id",
        },
        "property": Object {
          Symbol(refId): "gen_id",
        },
      },
      "property": Object {
        "id": "gen_id",
        "name": "Indiana Avenue",
        "owner": null,
        "price": 220,
        "rent": 100,
      },
    },
    Object {
      "node": Object {
        "id": "gen_id",
        "next": Object {
          Symbol(refId): "gen_id",
        },
        "prev": Object {
          Symbol(refId): "gen_id",
        },
        "property": Object {
          Symbol(refId): "gen_id",
        },
      },
      "property": Object {
        "id": "gen_id",
        "name": "Illinois Avenue",
        "owner": null,
        "price": 240,
        "rent": 100,
      },
    },
    Object {
      "node": Object {
        "id": "gen_id",
        "next": Object {
          Symbol(refId): "gen_id",
        },
        "prev": Object {
          Symbol(refId): "gen_id",
        },
        "property": Object {
          Symbol(refId): "gen_id",
        },
      },
      "property": Object {
        "id": "gen_id",
        "name": "Atlantic Avenue",
        "owner": null,
        "price": 260,
        "rent": 100,
      },
    },
    Object {
      "node": Object {
        "id": "gen_id",
        "next": Object {
          Symbol(refId): "gen_id",
        },
        "prev": Object {
          Symbol(refId): "gen_id",
        },
        "property": Object {
          Symbol(refId): "gen_id",
        },
      },
      "property": Object {
        "id": "gen_id",
        "name": "Ventnor Avenue",
        "owner": null,
        "price": 260,
        "rent": 100,
      },
    },
    Object {
      "node": Object {
        "id": "gen_id",
        "next": Object {
          Symbol(refId): "gen_id",
        },
        "prev": Object {
          Symbol(refId): "gen_id",
        },
        "property": Object {
          Symbol(refId): "gen_id",
        },
      },
      "property": Object {
        "id": "gen_id",
        "name": "Marvin Gardens",
        "owner": null,
        "price": 280,
        "rent": 100,
      },
    },
    Object {
      "node": Object {
        "id": "gen_id",
        "next": Object {
          Symbol(refId): "gen_id",
        },
        "prev": Object {
          Symbol(refId): "gen_id",
        },
        "property": Object {
          Symbol(refId): "gen_id",
        },
      },
      "property": Object {
        "id": "gen_id",
        "name": "Parcific Avenue",
        "owner": null,
        "price": 300,
        "rent": 100,
      },
    },
    Object {
      "node": Object {
        "id": "gen_id",
        "next": Object {
          Symbol(refId): "gen_id",
        },
        "prev": Object {
          Symbol(refId): "gen_id",
        },
        "property": Object {
          Symbol(refId): "gen_id",
        },
      },
      "property": Object {
        "id": "gen_id",
        "name": "North Carolina Avenue",
        "owner": null,
        "price": 300,
        "rent": 100,
      },
    },
  ],
  "players": Array [
    Object {
      "cash": 1000,
      "id": "player1",
      "owning": Array [],
      "position": Object {
        Symbol(refId): "gen_id",
      },
      "rolled": false,
    },
    Object {
      "cash": 1000,
      "id": "player2",
      "owning": Array [],
      "position": Object {
        Symbol(refId): "gen_id",
      },
      "rolled": false,
    },
  ],
}
`;

exports[`initial state 2`] = `"{\\"currentPlayer\\":\\"player1\\",\\"players\\":[{\\"id\\":\\"player1\\",\\"position\\":{},\\"owning\\":[],\\"cash\\":1000,\\"rolled\\":false},{\\"id\\":\\"player2\\",\\"position\\":{},\\"owning\\":[],\\"cash\\":1000,\\"rolled\\":false}],\\"map\\":[{\\"node\\":{\\"id\\":\\"gen_id\\",\\"prev\\":{},\\"next\\":{},\\"property\\":null},\\"property\\":null},{\\"node\\":{\\"id\\":\\"gen_id\\",\\"prev\\":{},\\"next\\":{},\\"property\\":{}},\\"property\\":{\\"id\\":\\"gen_id\\",\\"name\\":\\"Medierranean Avenue\\",\\"price\\":60,\\"rent\\":100,\\"owner\\":null}},{\\"node\\":{\\"id\\":\\"gen_id\\",\\"prev\\":{},\\"next\\":{},\\"property\\":{}},\\"property\\":{\\"id\\":\\"gen_id\\",\\"name\\":\\"Baltic Avenue\\",\\"price\\":60,\\"rent\\":100,\\"owner\\":null}},{\\"node\\":{\\"id\\":\\"gen_id\\",\\"prev\\":{},\\"next\\":{},\\"property\\":{}},\\"property\\":{\\"id\\":\\"gen_id\\",\\"name\\":\\"Oriental Avenue\\",\\"price\\":100,\\"rent\\":100,\\"owner\\":null}},{\\"node\\":{\\"id\\":\\"gen_id\\",\\"prev\\":{},\\"next\\":{},\\"property\\":{}},\\"property\\":{\\"id\\":\\"gen_id\\",\\"name\\":\\"Vermont Avenue\\",\\"price\\":100,\\"rent\\":100,\\"owner\\":null}},{\\"node\\":{\\"id\\":\\"gen_id\\",\\"prev\\":{},\\"next\\":{},\\"property\\":{}},\\"property\\":{\\"id\\":\\"gen_id\\",\\"name\\":\\"Connecticut Avenue\\",\\"price\\":120,\\"rent\\":100,\\"owner\\":null}},{\\"node\\":{\\"id\\":\\"gen_id\\",\\"prev\\":{},\\"next\\":{},\\"property\\":{}},\\"property\\":{\\"id\\":\\"gen_id\\",\\"name\\":\\"St. Charles Place\\",\\"price\\":140,\\"rent\\":100,\\"owner\\":null}},{\\"node\\":{\\"id\\":\\"gen_id\\",\\"prev\\":{},\\"next\\":{},\\"property\\":{}},\\"property\\":{\\"id\\":\\"gen_id\\",\\"name\\":\\"States Avenue\\",\\"price\\":140,\\"rent\\":100,\\"owner\\":null}},{\\"node\\":{\\"id\\":\\"gen_id\\",\\"prev\\":{},\\"next\\":{},\\"property\\":{}},\\"property\\":{\\"id\\":\\"gen_id\\",\\"name\\":\\"Virginia Avenue\\",\\"price\\":160,\\"rent\\":100,\\"owner\\":null}},{\\"node\\":{\\"id\\":\\"gen_id\\",\\"prev\\":{},\\"next\\":{},\\"property\\":{}},\\"property\\":{\\"id\\":\\"gen_id\\",\\"name\\":\\"St.James Place\\",\\"price\\":180,\\"rent\\":100,\\"owner\\":null}},{\\"node\\":{\\"id\\":\\"gen_id\\",\\"prev\\":{},\\"next\\":{},\\"property\\":{}},\\"property\\":{\\"id\\":\\"gen_id\\",\\"name\\":\\"Tennessee Avenue\\",\\"price\\":180,\\"rent\\":100,\\"owner\\":null}},{\\"node\\":{\\"id\\":\\"gen_id\\",\\"prev\\":{},\\"next\\":{},\\"property\\":{}},\\"property\\":{\\"id\\":\\"gen_id\\",\\"name\\":\\"New York Avenue\\",\\"price\\":200,\\"rent\\":100,\\"owner\\":null}},{\\"node\\":{\\"id\\":\\"gen_id\\",\\"prev\\":{},\\"next\\":{},\\"property\\":{}},\\"property\\":{\\"id\\":\\"gen_id\\",\\"name\\":\\"Kentucky Avenue\\",\\"price\\":220,\\"rent\\":100,\\"owner\\":null}},{\\"node\\":{\\"id\\":\\"gen_id\\",\\"prev\\":{},\\"next\\":{},\\"property\\":{}},\\"property\\":{\\"id\\":\\"gen_id\\",\\"name\\":\\"Indiana Avenue\\",\\"price\\":220,\\"rent\\":100,\\"owner\\":null}},{\\"node\\":{\\"id\\":\\"gen_id\\",\\"prev\\":{},\\"next\\":{},\\"property\\":{}},\\"property\\":{\\"id\\":\\"gen_id\\",\\"name\\":\\"Illinois Avenue\\",\\"price\\":240,\\"rent\\":100,\\"owner\\":null}},{\\"node\\":{\\"id\\":\\"gen_id\\",\\"prev\\":{},\\"next\\":{},\\"property\\":{}},\\"property\\":{\\"id\\":\\"gen_id\\",\\"name\\":\\"Atlantic Avenue\\",\\"price\\":260,\\"rent\\":100,\\"owner\\":null}},{\\"node\\":{\\"id\\":\\"gen_id\\",\\"prev\\":{},\\"next\\":{},\\"property\\":{}},\\"property\\":{\\"id\\":\\"gen_id\\",\\"name\\":\\"Ventnor Avenue\\",\\"price\\":260,\\"rent\\":100,\\"owner\\":null}},{\\"node\\":{\\"id\\":\\"gen_id\\",\\"prev\\":{},\\"next\\":{},\\"property\\":{}},\\"property\\":{\\"id\\":\\"gen_id\\",\\"name\\":\\"Marvin Gardens\\",\\"price\\":280,\\"rent\\":100,\\"owner\\":null}},{\\"node\\":{\\"id\\":\\"gen_id\\",\\"prev\\":{},\\"next\\":{},\\"property\\":{}},\\"property\\":{\\"id\\":\\"gen_id\\",\\"name\\":\\"Parcific Avenue\\",\\"price\\":300,\\"rent\\":100,\\"owner\\":null}},{\\"node\\":{\\"id\\":\\"gen_id\\",\\"prev\\":{},\\"next\\":{},\\"property\\":{}},\\"property\\":{\\"id\\":\\"gen_id\\",\\"name\\":\\"North Carolina Avenue\\",\\"price\\":300,\\"rent\\":100,\\"owner\\":null}}]}"`;
